docs/docs/typelevel.md

```diff
@@ -0,0 +1,532 @@
+
+    inline def toInt[N <: Nat]: Int = ~{
+      case N =:= Z => '0
+      case N =:= S[type N1] => '(1 + toInt[N1])
+    }
+
+A query conditional is written ` { <cases> }` where each case is of the form
+
+    case <type-pattern> <guard> => <expr>
+
+`<guard>` is either empty or of the form `if <expr>`. Some examples of legal cases are
+
+    case Ord[X] => e1
+    case X <:< List[Y] if n == 0 => e2
+    case _ if n < 0 => e3
+
+Query conditionals may only appear in spliced code. That sets them
+apart from blocks with normal case clauses which define partial
+functions and therefore can only appear in normal code.  Every query
+conditional is evaluated by the macro interpreter. The macro
+interpreter evaluates each case of a query conditional by performing
+an implicit search for the given type pattern. If the search succeeds,
+it continues with evaluating the guard, if there is one, followed by
+the right-hand side, otherwise it continues with the next case.
+
+Type patterns may bind type variables. For instance,
+`N =:= S[type N1]` binds the type variable `N1`. Bound type variables are
+instantiated by the implicit search. They are visible in the rest of
+the case.
+
+In general, a type name in an implicit condition is treated as a bound
+variable if it is prefixed by `type`.
+
+_Aside:_ We should introduce the same convention for types in normal
+patterns. Currently type identifiers in patterns are treated as bound
+if they start with a lower-case letter and as free otherwise. This
+mirrors the convention for names, but is not very visible. Requiring
+`type` prefixes instead would make programs clearer.
+
+In last clause of the code above, the macro interpreter would perform
+an implicit search for the type `N =:= S[type N1]` where `N` is the
+actual type passed to `toInt` and `N1` is a type variable bound by the
+search.  If the search succeeds, it returns one plus the result of
+evaluating `toInt[N1]`. If the search fails, it results in compilation
+failure.
+
+The result of `toInt[S[S[Z]]]` would hence be the integer `1 + (1 +
+0)`, which can be constant-folded to `2`. Alternatvely, we could drop
+the use of quotes and rely on lifting to go from `Int` to
+`Expr[Int]`. So the following would also be correct and would return a
+constant without relying on constant folding:
+
+    inline def toInt[N <: Nat]: Int = ~{
+      case N =: Z => 0
+      case N =:= S[type N1] => 1 + toInt[N1]
+    }
+
+### Example: HLists
+
+Consider the standard definition of `HList`s:
+
+    class HList
+
+    case class HCons[X, Xs <: HList](hd: X, tl: Xs) extends HList
+    case object HNil extends HList
+
+A type-level length function on HLists can be written as follows:
+
+    type Length[Xs <: HList] <: Nat = {
+      case Xs =:= HNil => Z
+      case Xs =:= HCons[type X, type Xs1] => S[Length[Xs1]]
+    }
+
+If we don't want to introduce the type variable `X`, which is unused, we could
+also formulate the type pattern with `<:<` instead of `=:=`:
```

Comment:
Why can’t we just use `_` and `<:<` together, and write `Xs =:= HCons[_, type Xs1]`? The semantics would be a conventional match, followed by throwing away `X`, per the usual semantics of `_`. No subtyping should be necessary.

Subtyping might instead be needed if `Xs` were a *subtype* of `HCons[X, Xs1]`. It’d be good to check how `=:=` and subtyping interact and if the interaction is acceptable in a wider range of use cases.

Reply:
We could probably introduce

    Xs =:= HCons[type _, type Xs1]

if we want to be fancy. But

    Xs =:= HCons[_, type Xs1]

already means something else: That the type `Xs` is equal to the type `HCons[_, T]` for some `T`, where `_` is a wildcard. If we write this out in terms of DOT it would be something like

    HCons { type Hd; type Tl = T }

Reply:
Good point, but doesn’t `_` in `Xs <:< HCons[_, type Xs1]` also refer to a wildcard? Or what do I miss?

Reply:
`HCons[Int, Nil] <:< HCons[_, Nil]` but `=:=` between the types does not hold.

Reply:
Ah the existing example is already using the wildcard semantics, not the throw-away semantics! Hadn’t realized. For later: a syntax less ambiguous for readers might be nice.